cdap-api/src/main/java/co/cask/cdap/api/plugin/PluginClass.java

```diff
@@ -62,13 +64,21 @@ public PluginClass(String type, String name, String description, String classNam
     this.description = description;
     this.className = className;
     this.configFieldName = configfieldName;
-    this.properties = properties;
-    this.endpoints = endpoints;
+    this.properties = Collections.unmodifiableMap(new HashMap<>(properties));
+    this.endpoints = Collections.unmodifiableSet(new HashSet<>(endpoints));
+    this.requirements = requirements == null ? Collections.emptySet() :
```

Comment:
I don't see `requirements` is annotated with `Nullable`. When will it be null?

Reply:
Someone might call with null but we always want to ensure that we have empty requirement fixed it throw exception on null and also internally call with empty set

Reply:
Who is someone? Why can't we might sure it won't be null, similar to properties and endpoints?

Reply:
Any who calls PluginClass constructor. In comment I mentioned I updated it to throw exception on null. I don't see endpoint has a check. I will add one for that.